File: ParserLib/src/grammar/production.rs
use crate::grammar::symbol::Symbol;

#[derive(Debug,Clone,Eq, PartialEq)]
pub struct Production
{
    pub left:Symbol,
    pub right:Vec<Symbol>
}
impl Production
{
    pub fn new(left:Symbol,right:Vec<Symbol>)->Self
    {
        Production{left,right}
    }
    pub fn from_str(left:String,right:Vec<Symbol>)->Self
    {
        Production{left:Symbol::non_terminal_with(left),right}
    }
    pub fn print(&self)
    {
        print!("{} ->",self.left.value);
        for i in self.right.iter()
        {
            print!(" {}", i.value);
        }
        print!("\n");
    }
}

File: ParserLib/Grammar/Production.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ParserLib.Grammar;

namespace ParserLib.Grammar;

public readonly struct Production:IEquatable<Production>
{
    public Symbol Left { get; init; }
    public List<Symbol> Right { get; init; }

    public Production(string left, List<Symbol> right)
    {
        Left = Symbol.NonTerminal(left);
        Right= right;
    }
    public Production(Symbol left, List<Symbol> right)
    {
        Left = left;
        Right = right;
    }

    public void Print()
    {
        Console.WriteLine(ToString());
    }

    public override string ToString()
    {
        return $"{Left.Value} -> {string.Join(" ", Right.Select((x => x.Value)))}";
    }

    public bool Equals(Production other)
    {
        if (!Left.Equals(other.Left)||Right.Count!=other.Right.Count)
            return false;

        for (int i = 0; i < Right.Count; i++)
        {
            if(Right[i]!=other.Right[i])
                return false;
        }
        return true;
    }

    public override bool Equals(object? obj)
    {
        return obj is Production other && Equals(other);
    }
    public static bool operator !=(Production lhs, Production rhs)
    {
        return !Equals(lhs, rhs);
    }
    public static bool operator ==(Production lhs, Production rhs)
    {
        return Equals(lhs, rhs);
    }
    public override int GetHashCode()
    {
        return HashCode.Combine(Left, Right);
    }
}